SpatialGDK/Source/SpatialGDKServices/Private/SSpatialOutputLog.cpp

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SSpatialOutputLog.h"
+
+#include "Async/Async.h"
+#include "DirectoryWatcherModule.h"
+#include "Editor.h"
+#include "Misc/CoreDelegates.h"
+#include "Misc/FileHelper.h"
+#include "Modules/ModuleManager.h"
+#include "SlateOptMacros.h"
+#include "SpatialGDKServicesModule.h"
+
+#define LOCTEXT_NAMESPACE "SSpatialOutputLog"
+
+DEFINE_LOG_CATEGORY(LogSpatialOutputLog);
+
+static const FString LocalDeploymentLogsDir(FSpatialGDKServicesModule::GetSpatialOSDirectory(TEXT("logs/localdeployment")));
+static const FString LaunchLogFilename(TEXT("launch.log"));
+
+void FArchiveLogFileReader::UpdateFileSize()
+{
+	Size = IFileManager::Get().GetStatData(*Filename).FileSize;
+}
+
+TUniquePtr<FArchiveLogFileReader> SSpatialOutputLog::CreateLogFileReader(const TCHAR* InFilename, uint32 Flags, uint32 BufferSize)
+{
+	IFileHandle* Handle = FPlatformFileManager::Get().GetPlatformFile().OpenRead(InFilename, !!(Flags & FILEREAD_AllowWrite));
+	if (!Handle)
+	{
+		if (Flags & FILEREAD_NoFail)
+		{
+			UE_LOG(LogSpatialOutputLog, Error, TEXT("Failed to read file: %s"), InFilename);
+		}
+
+		return nullptr;
+	}
+
+	return TUniquePtr<FArchiveLogFileReader>(new FArchiveLogFileReader(Handle, InFilename, Handle->Size(), BufferSize));
+}
+
+BEGIN_SLATE_FUNCTION_BUILD_OPTIMIZATION
+void SSpatialOutputLog::Construct(const FArguments& InArgs)
+{
+	SOutputLog::Construct(SOutputLog::FArguments());
+
+	// Remove ourselves as the constructor of our parent (SOutputLog) added 'this' as a remote output device.
+	GLog->RemoveOutputDevice(this);
+
+	LogReader = nullptr;
+
+	StartUpLogDirectoryWatcher(LocalDeploymentLogsDir);
+
+	// Set the LogReader to the latest launch.log if we can.
+	ReadLatestLogFile();
+}
+END_SLATE_FUNCTION_BUILD_OPTIMIZATION
+
+void SSpatialOutputLog::ReadLatestLogFile()
+{
+	FString NewestLogDir;
+	FDateTime NewestLogDirTime;
+
+	// Go through all log directories in the spatial logs and find the most recently created (if one exists) and print the log file to the Spatial Output.
+	bool GetNewestLogDir = IFileManager::Get().IterateDirectoryStat(*LocalDeploymentLogsDir, [&NewestLogDir, &NewestLogDirTime](const TCHAR* FileName, const FFileStatData& FileStats) {
+
+		if (FileStats.bIsDirectory)
+		{
+			if (FileStats.CreationTime > NewestLogDirTime)
+			{
+				NewestLogDir = FString(FileName);
+				NewestLogDirTime = FileStats.CreationTime;
+			}
+		}
+
+		return true;
+	});
+
+	if (GetNewestLogDir)
+	{
+		StartPollingLogFile(FPaths::Combine(NewestLogDir, LaunchLogFilename));
+	}
+}
+
+SSpatialOutputLog::~SSpatialOutputLog()
+{
+	CloseLogReader();
+
+	ShutdownLogDirectoryWatcher(LocalDeploymentLogsDir);
+
+	FCoreDelegates::OnHandleSystemError.RemoveAll(this);
+}
+
+void SSpatialOutputLog::OnCrash()
+{
+	CloseLogReader();
+
+	ShutdownLogDirectoryWatcher(LocalDeploymentLogsDir);
+}
+
+void SSpatialOutputLog::StartUpLogDirectoryWatcher(FString LogDirectory)
```

Comment:
Can we put async in the names of methods that just start perform some task at a later time?

Reply:
This can be passed by const ref